scripts: kconfig: normalize paths to avoid duplicate

hakonfam opened the pull request:
There is a bug where non-normalized paths can introduce multiple entries
for the same file. E.g. '/1/2/../2.file' and '/1/2/3/../../2.file' would
both be listed, and end in a ninja error because multiple targets
generate the same file.

This commit fixes this issue by normalizing the paths.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

SebastianBoe commented on scripts/kconfig/kconfig.py:
```diff
@@ -195,20 +197,16 @@ def write_kconfig_filenames(paths, root_path, output_file_path):
 
     # Remove duplicates
-    paths_uniq = set(paths)
+    paths_uniq = set([normpath(abspath(join(root_path, path))) for path in paths])
```
This does more than remove duplicates so the comment needs updating. Could possibly be removed.

hakonfam replied:
I agree, the name of the variable tells it all.

SebastianBoe replied:
The fact that we normalize paths, and what that means, should be commented somewhere.

hakonfam committed:
scripts: kconfig: normalize paths to avoid duplicate

There is a bug where non-normalized paths can introduce multiple entries
for the same file. E.g. '/1/2/../2.file' and '/1/2/3/../../2.file' would
both be listed, and end in a ninja error because multiple targets
generate the same file.

This commit fixes this issue by normalizing the paths.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

ulfalizer commented:
@hakonfam 
Could you give an example of a file that gets included twice under different paths, and what the paths are?

Trying to understand how this happens.

ulfalizer commented on scripts/kconfig/kconfig.py:
```diff
@@ -184,31 +185,26 @@ def promptless(sym):
 
-    # Remove duplicates
-    paths_uniq = set(paths)
+    paths_uniq = sorted(set([normpath(abspath(join(root_path, path))) for path in paths]))
```
Can use a generator to get rid of a temporary list (pair of `[]`). Think `realpath()` would work here too.

```python3
paths_uniq = sorted(set(realpath(join(root_path, path)) for path in paths))
```

ulfalizer replied:
Can be simplified a bit more with a set comprehension:

```python
paths_uniq = sorted({realpath(join(root_path, path)) for path in paths})
```

ulfalizer commented on scripts/kconfig/kconfig.py:
```diff
@@ -184,31 +185,26 @@ def promptless(sym):
-    # The written list should also have absolute paths instead of
-    # relative paths, and it should not have duplicates.
+    # The written list has sorted normalized absolute paths, and it does not have duplicates.
+    # The list is sorted to be deterministic.
```
Could explain here why the files need to be `realpath()`'d. I'm curious too. :)

hakonfam commented:
> @hakonfam
> Could you give an example of a file that gets included twice under different paths, and what the paths are?

I don't have an upstream example, but this would trigger the bug:
```
# file ~/r/zephyr/subsys/foo/foo.kconf 
rsource "../templates/something.template"

# file ~/somewhere/else/bar.kconf
source "${ZEPHYR_BASE}/subsys/templates/something.template"
```

ulfalizer commented:
Ah, thought it was triggered by something upstream. No problem for me removing duplicates if the CMake stuff chokes on it at least.

ioannisg commented:
@nashif @SebastianBoe do we need to have this in 2.0, or is, rather, a lower-prio issue that can be merged after release?

SebastianBoe commented:
This should be in 2.0.

ioannisg commented:
@nashif could you please review this patch?
